rclcpp/include/rclcpp/mapped_ring_buffer.hpp

```diff
@@ -0,0 +1,205 @@
+ * This class must have a positive, non-zero size.
+ * This class cannot be resized nor can it reserve additional space after construction.
+ * This class is not CopyConstructable nor CopyAssignable.
+ *
+ * The key's are not guaranteed to be unique because push_and_replace does not
+ * check for colliding keys.
+ * It is up to the user to only use unique keys.
+ * A side effect of this is that when get_copy_at_key or pop_at_key are called,
+ * they return the first encountered instance of the key.
+ * But iteration does not begin with the ring buffer's head, and therefore
+ * there is no guarantee on which value is returned if a key is used multiple
+ * times.
+ */
+template<typename T>
+class MappedRingBuffer : public MappedRingBufferBase
+{
+public:
+  RCLCPP_SMART_PTR_DEFINITIONS(MappedRingBuffer<T>);
+
+  /// Constructor.
+  /* The constructor will allocate memory while reserving space.
+   *
+   * /param size size of the ring buffer; must be positive and non-zero.
+   */
+  MappedRingBuffer(size_t size) : elements_(size), head_(0)
+  {
+    if (size == 0) {
+      throw std::invalid_argument("size must be a positive, non-zero value");
+    }
+  }
+  virtual ~MappedRingBuffer() {}
+
+  /// Return a copy of the value stored in the ring buffer at the given key.
+  /* The key is matched if an element in the ring buffer has a matching key.
+   * This method will allocate in order to return a copy.
+   *
+   * The key is not guaranteed to be unique, see the class docs for more.
+   *
+   * /param key the key associated with the stored value
+   * /param value if the key is found, the value is stored in this parameter
+   */
+  void
+  get_copy_at_key(uint64_t key, std::unique_ptr<T> & value)
+  {
+    auto it = get_iterator_of_key(key);
+    value = nullptr;
+    if (it != elements_.end() && it->in_use) {
+      value = std::unique_ptr<T>(new T(*it->value));
+    }
+  }
+
+  /// Return ownership of the value stored in the ring buffer, leaving a copy.
+  /* The key is matched if an element in the ring bufer has a matching key.
+   * This method will allocate in order to store a copy.
+   *
+   * The key is not guaranteed to be unique, see the class docs for more.
+   *
+   * The ownership of the currently stored object is returned, but a copy is
+   * made and stored in its place.
+   * This means that multiple calls to this function for a particular element
+   * will result in returning the copied and stored object not the original.
+   * This also means that later calls to pop_at_key will not return the
+   * originally stored object, since it was returned by the first call to this
+   * method.
+   *
+   * /param key the key associated with the stored value
+   * /param value if the key is found, the value is stored in this parameter
+   */
+  void
+  get_ownership_at_key(uint64_t key, std::unique_ptr<T> & value)
+  {
+    auto it = get_iterator_of_key(key);
+    value = nullptr;
```

Comment:
It's up to the user, but I added a note to each of the affected methods: 435591d